src/__tests__/use-collection-expadable-rows.test.tsx

```diff
@@ -0,0 +1,342 @@
+    const result = useCollection(items, {
+      expandableRows: {
+        getId,
+        getParentId,
+        defaultExpandedItems: allItems,
+      },
+    });
+    return <Demo {...result} />;
+  }
+
+  const { rerender, getExpandedItems } = render(<App items={allItems} />);
+  expect(getExpandedItems()).toEqual(['a', 'b', 'c', 'd', 'e', 'f']);
+
+  rerender(<App items={[allItems[0], allItems[2], allItems[3], allItems[4], allItems[5]]} />);
+  expect(getExpandedItems()).toEqual(['a', 'c', 'd', 'e', 'f']);
+
+  rerender(<App items={allItems} />);
+  expect(getExpandedItems()).toEqual(['a', 'c', 'd', 'e', 'f']);
+});
+
+test('expanded rows with text filtering', () => {
+  const deepTreeItemsWithValues: (Item & { value?: string })[] = [...deepTreeItems];
+  deepTreeItemsWithValues.find(item => item.id === 'a.1')!.value = 'match';
+
+  function App() {
+    const result = useCollection(deepTreeItemsWithValues, {
+      expandableRows: {
+        getId,
+        getParentId: getDeepTreeParentId,
+        defaultExpandedItems: deepTreeItemsWithValues,
+      },
+      filtering: {
+        defaultFilteringText: 'match',
+      },
+    });
+    return <Demo {...result} />;
+  }
+  const { getVisibleItems } = render(<App />);
+
+  expect(getVisibleItems()).toEqual(['a', 'a.1']);
+});
+
+test('expanded rows with property filtering', () => {
+  const deepTreeItemsWithValues: (Item & { value?: string })[] = [...deepTreeItems];
+  deepTreeItemsWithValues.find(item => item.id === 'a.1')!.value = 'match';
+
+  function App() {
+    const result = useCollection(deepTreeItemsWithValues, {
+      expandableRows: {
+        getId,
+        getParentId: getDeepTreeParentId,
+        defaultExpandedItems: deepTreeItemsWithValues,
+      },
+      propertyFiltering: {
+        filteringProperties: [{ key: 'value', operators: ['='], propertyLabel: '', groupValuesLabel: '' }],
+        defaultQuery: { tokens: [{ propertyKey: 'value', operator: '=', value: 'match' }], operation: 'and' },
+      },
+    });
+    return <Demo {...result} />;
+  }
+  const { getVisibleItems } = render(<App />);
+
+  expect(getVisibleItems()).toEqual(['a', 'a.1']);
+});
+
+test('expanded rows with pagination', () => {
+  function App({ pageSize }: { pageSize: number }) {
+    const result = useCollection(treeItems, {
+      pagination: { pageSize },
+      expandableRows: {
+        getId,
+        getParentId: getTreeParentId,
+        defaultExpandedItems: treeItems,
+      },
+    });
+    return <Demo {...result} />;
+  }
+  const { rerender, getVisibleItems } = render(<App pageSize={10} />);
+
+  expect(getVisibleItems()).toEqual(treeItems.map(item => item.id));
+
+  rerender(<App pageSize={3} />);
+  expect(getVisibleItems()).toEqual(treeItems.map(item => item.id).slice(0, 14));
+});
+
+test('expanded rows with sorting', () => {
+  const shuffledDeepTreeItems: Item[] = deepTreeItems
+    .map(item => ({ ...item, value: Math.random() }))
```

Comment:
Not sure about having randomization in tests, I think it's better practice to make them predictable. I would rather come up with a hard-coded disordered list to test against. If several disorder cases are considered relevant, then have one test for each.

Reply:
That is inspired by property-based testing. The test ensures literally that the original order of items is of no importance - the result must always be sorted. Considered the length of the array there is a way to small chance that the test works by accident - when the random sorting would produce the original array.